src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,208 @@
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space distance r
+via cubic spline interpolation on the real-space mesh.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so we must divide by `r`.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
+    ir_cut = length(psp.projs[l+1][i])
+    linear_interpolation((psp.rgrid[1:ir_cut],), psp.projs[l+1][i])(r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l in at k-space distance q.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so the integrand has `r` instead of `r^2`
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    if iszero(q)
+        if iszero(l)
+            s = one(T)
+        else
+            s = zero(T)
+        end
+    else
+        l_val = Val(l)
+        s = psp_projector_fouier_inner_loop(l_val, psp, i, l, q)
+        s = 2T(π) * s  # 2π because Ry -> Ha included
+    end
+    return s
+end
+
+function psp_projector_fouier_inner_loop(order::Val, psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    N = length(psp.rprojsdr[l+1][i])
+    s = zero(T)
+    @inbounds for ir = 1:N
+        s += sphericalbesselj(order, q * psp.rgrid[ir]) * psp.rprojsdr[l+1][i][ir]
+    end
+    return s
+end
+
+"""
+     eval_psp_local_real(psp::PspUpf, r::Number)
+
+ Evaluate the local potential at real-space distance `r` via linear interpolation on the
+ real-space mesh.
+ """
+function eval_psp_local_real(psp::PspUpf, r::T) where {T <: Real}
+    linear_interpolation((psp.rgrid,), psp.vloc)(r) / 2  # Ry -> Ha
+end
+
+"""
+    eval_psp_local_fourier(psp::PspUpf, q::Number)
+
+Evaluate the local potential at fourier-space distance `q` using a Coulomb correction term
+`-Z/r`.
+"""
+function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
+    Z = 2 * psp.Zion  # for some strange reason
+    N = length(psp.rvlocpzdr)
+    s = zero(T)
+    @inbounds @fastmath for ir = 1:N
+        # sin(qr) / qr * (V(r) - -Z/r) r^2 dr
+        s += sin(q * psp.rgrid[ir]) * psp.rvlocpzdr[ir]
+    end
+    2T(π) * (s - Z/q) / q  # 2π because Ry -> Ha included
+end
+
+"""
+    eval_psp_energy_correction(T::Type, psp::PspUpf, n_electrons::Number)
+
+Evaluate the pseudopotential energy correction, i.e. the limit of the Fourier-transform of the
+local potential as q -> 0.
+"""
+function eval_psp_energy_correction(T, psp::PspUpf, n_electrons)
+    N = length(psp.rvlocpzdr)
+    s = zero(T)
+    @inbounds for ir = 1:N
+        # (V(r) - -Z/r) r^2 dr
+        s += psp.rgrid[ir] * psp.rvlocpzdr[ir]
+    end
+    2T(π) * n_electrons * s  # 2π because Ry -> Ha included
+end
+
+@inline @fastmath function sphericalbesselj(order::Val{0}, x)
```

Comment:
I would move the `@fastmath` to the loop (next to the `@inbounds` ... or maybe even drop it. In my tests it did nothing.

Reply:
I tried this, and it seems like `@fastmath` didn't go through the function barrier. In my testing, it did make some difference, how big I'm not entirely sure.

Reply:
I'm really not a fan of fastmath. It can lead to spurious errors down the road ([Blog article](https://simonbyrne.github.io/notes/fastmath/)). Unless it brings some real speedup, I would avoid having it.